SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyTracer.h

```diff
@@ -0,0 +1,112 @@
+	// 2. Create and download a service-account certificate
+	// 3. Set GOOGLE_APPLICATION_CREDENTIALS to certificate path
+	// 4. Set GRPC_DEFAULT_SSL_ROOTS_FILE_PATH to your `roots.pem` gRPC path
+	//
+	// Usage:
+	// 1. Register your Google's project id with `RegisterProject`
+	// 2. Start a latency trace using `BeginLatencyTrace` tagging it against an Actor's RPC
+	// 3. During the execution of the tagged RPC either;
+	//		- continue the trace using `ContinueLatencyTrace`, again tagging it against another Actor's RPC
+	//		- or end the trace using `EndLatencyTrace`
+	//
+	//////////////////////////////////////////////////////////////////////////
+
+	// Front-end exposed, allows users to register, start, continue, and end traces
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static void RegisterProject(UObject* WorldContextObject, const FString& ProjectId);
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static bool BeginLatencyTrace(UObject* WorldContextObject, const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static bool ContinueLatencyTrace(UObject* WorldContextObject, const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+
+	UFUNCTION(BlueprintCallable, Category = "SpatialOS", meta = (WorldContext = "WorldContextObject"))
+	static bool EndLatencyTrace(UObject* WorldContextObject);
+
+	static const TraceKey ActiveTraceKey = 0;
+	static const TraceKey InvalidTraceKey = -1;
+
+#if TRACE_LIB_ACTIVE
+
+	// Internal GDK usage, shouldn't be used by game code
+	static USpatialLatencyTracer* GetTracer(UObject* WorldContextObject);
+
+	bool IsValidKey(TraceKey Key);
+	TraceKey GetTraceKey(const UObject* Obj, const UFunction* Function);
+
+	void WriteToLatencyTrace(const TraceKey Key, const FString& TraceDesc);
+	void EndLatencyTrace(const TraceKey Key, const FString& TraceDesc);
+
+	void WriteTraceToSchemaObject(const TraceKey Key, Schema_Object* Obj, const Schema_FieldId FieldId);
+	TraceKey ReadTraceFromSchemaObject(Schema_Object* Obj, const Schema_FieldId FieldId);
+
+	void SetWorkerId(const FString& NewWorkerId) { WorkerId = NewWorkerId; }
+
+private:
+
+	using ActorFuncKey = TPair<const AActor*, const UFunction*>;
+	using TraceSpan = improbable::trace::Span;
+
+	bool BeginLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+	bool ContinueLatencyTrace_Internal(const AActor* Actor, const FString& FunctionName, const FString& TraceDesc);
+	bool EndLatencyTrace_Internal();
+
+	TraceKey CreateNewTraceEntry(const AActor* Actor, const FString& FunctionName);
+	TraceSpan* GetActiveTrace();
+
+	void WriteKeyFrameToTrace(const TraceSpan* Trace, const FString& TraceDesc);
+	FString FormatMessage(const FString& Message) const;
+
+	FString WorkerId;
```

Comment:
Should we worry about WorkerId not being initialized for some time, since there's a gap between a call to `Init` and `SetWorkerId`?

Reply:
Set a default.